libtenzir/builtins/operators/time.cpp

```diff
@@ -0,0 +1,607 @@
+    -> std::tuple<
+      std::optional<std::chrono::time_point<Clock, std::chrono::seconds>>,
+      std::optional<long>, std::optional<std::string>> {
+    auto tm_value = to_tm(diag);
+    if (not tm_value)
+      return {std::nullopt, std::nullopt, std::nullopt};
+    auto time = std::chrono::seconds{tm_value->tm_sec + 60 * tm_value->tm_min
+                                     + 60 * 60 * tm_value->tm_hour};
+    if (time >= std::chrono::seconds{86400} || time < std::chrono::seconds{0}) {
+      diagnostic::error("invalid time").note("value: {}", time).emit(diag);
+      return {std::nullopt, std::nullopt, std::nullopt};
+    }
+    auto date = date::year{tm_value->tm_year + 1900}
+                / date::month{static_cast<unsigned>(tm_value->tm_mon + 1)}
+                / date::day{static_cast<unsigned>(tm_value->tm_mday)};
+    date += date::months{0};
+    if (not date.ok()) {
+      diagnostic::error("invalid date")
+        .note("value: `{}-{}-{}`", static_cast<int>(date.year()),
+              static_cast<unsigned>(date.month()),
+              static_cast<unsigned>(date.day()))
+        .emit(diag);
+      return {std::nullopt, std::nullopt, std::nullopt};
+    }
+    auto datetime = std::chrono::time_point<Clock, std::chrono::seconds>{
+      ymd_to_days<Clock>(date) + time};
+    return {datetime, tm_gmtoff, tm_zone};
+  }
+
+  /// Returns a tuple of the time contained in `*this` in `date::local_seconds`
+  /// (local time_point), UTC offset, and timezone name.
+  ///
+  /// If `*this` is not complete, the first element of the returned tuple is
+  /// `std::nullopt`.
+  [[nodiscard]] auto to_local_time_point(diagnostic_handler& diag)
+    -> std::tuple<std::optional<date::local_seconds>, std::optional<long>,
+                  std::optional<std::string>> {
+    return to_clock_time_point<date::local_t>(diag);
+  }
+
+  /// Returns an object of type `std::chrono::sys_seconds` corresponding to
+  /// `*this`.
+  ///
+  /// If `is_complete()` is `false`, returns `std::nullopt`, and
+  /// emits an error to `diag`.
+  /// `*this` must be in UTC.
+  auto to_system_time_point(diagnostic_handler& diag)
+    -> std::optional<std::chrono::sys_seconds> {
+    auto [tp, _1, _2] = to_clock_time_point<std::chrono::system_clock>(diag);
+    if (not tp)
+      return std::nullopt;
+    TENZIR_ASSERT(is_utc());
+    return *tp;
+  }
+
+  void to_record(record_ref& builder) const {
+    auto add_field_if_set
+      = [&]<typename T>(std::string_view name, const std::optional<T>& val,
+                        std::function<T(T)> mod = std::identity{}) {
+          if (val)
+            builder.field(name, mod(*val));
+          else
+            builder.field(name, caf::none);
+        };
+    add_field_if_set("sec", tm_sec);
+    add_field_if_set("min", tm_min);
+    add_field_if_set("hour", tm_hour);
+    add_field_if_set("day", tm_mday);
+    add_field_if_set("month", tm_mon, {[](int mon) {
```

Comment:
Should we use
1. `sec`, `min`, and `mday`, or
2. `second`, `minute`, and `day`?

Not sure which is better, but right now we appear to be using a mix of spelled-out and what the underlying fields are called, which is a bit inconsistent.

Reply:
Definite no for `mday`. The field is called that in `std::tm` to distinguish it from `wday` (month-day as opposed to week-day), but we have no need for that. `mday` would also indicate that the valid range would be [0, 30], and not [1, 31].

As for `sec` and `min` vs. `second` and `minute`, I have no strong feelings either way.

Reply:
Then I would put my vote in for `second` and `minute`, such that all of these are complete words. 

(Also maybe `tz` -> `timezone`).